File: src/runtime/capabilities-resolver.ts
/**
 * @license
 * Copyright (c) 2020 Google Inc. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import {assert} from '../platform/assert-web.js';
import {Dictionary} from '../utils/lib-utils.js';
import {StorageKey} from './storage/storage-key.js';
import {Type} from '../types/lib-types.js';
import {ArcId} from './id.js';
import {Flags} from './flags.js';
import {Capabilities} from './capabilities.js';
import {ReferenceModeStorageKey} from './storage/reference-mode-storage-key.js';
import {StorageKeyFactory,
        FactorySelector,
        ContainerStorageKeyOptions,
        BackingStorageKeyOptions,
        SimpleCapabilitiesSelector} from './storage-key-factory.js';

export type CapabilitiesResolverOptions = Readonly<{
  arcId: ArcId;
  factories?: StorageKeyFactory[];
  selector? : FactorySelector;
}>;

export class CapabilitiesResolver {
  private readonly selector: FactorySelector;
  private readonly factories: Dictionary<StorageKeyFactory> = {};
  private readonly arcId: ArcId;

  constructor(options: CapabilitiesResolverOptions) {
    for (const factory of (options.factories || [])) {
      assert(!this.factories[factory.protocol], `Duplicated factory for '${factory.protocol}'.`);
      this.factories[factory.protocol] = factory;
    }
    this.selector = options.selector || new SimpleCapabilitiesSelector();
    this.arcId = options.arcId;
  }

  async createStorageKey(capabilities: Capabilities, type: Type, handleId: string): Promise<StorageKey> {
    const factory = this.selectStorageKeyFactory(capabilities, handleId);
    return this.createStorageKeyWithFactory(factory, type, handleId);
  }

  selectStorageKeyFactory(capabilities: Capabilities, handleId: string): StorageKeyFactory {
    const selectedFactories = Object.values(this.factories).filter(factory => {
      return factory.supports(capabilities);
    });

    if (selectedFactories.length === 0) {
      throw new Error(`Cannot create a suitable storage key for handle '${
        handleId}' with capabilities ${capabilities.toDebugString()}`);
    }
    return this.selector.select(selectedFactories);
  }

  private async createStorageKeyWithFactory(factory: StorageKeyFactory, type: Type, handleId: string): Promise<StorageKey> {
    const schema = type.getEntitySchema();
    const schemaHash = await schema.hash();
    const options = new ContainerStorageKeyOptions(this.arcId, schemaHash, schema.name);
    const containerKey = factory.create(options);
    const containerChildKey = containerKey.childKeyForHandle(handleId);
    if (!Flags.defaultReferenceMode) {
      return containerChildKey;
    }
    const containedType = type.getContainedType();
    if (type.isReference || (containedType && containedType.isReference)) {
      return containerChildKey;
    }
    const backingKey = factory.create(
      new BackingStorageKeyOptions(this.arcId, schemaHash, schema.name));
    // ReferenceModeStorageKeys in different drivers can cause problems with garbage collection.
    assert(backingKey.protocol === containerKey.protocol);
    return new ReferenceModeStorageKey(backingKey, containerChildKey);
  }
}

File: java/arcs/core/storage/CapabilitiesResolver.kt
/*
 * Copyright 2020 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.core.storage

import arcs.core.common.ArcId
import arcs.core.data.Capabilities
import arcs.core.data.ReferenceType
import arcs.core.data.toSchema
import arcs.core.storage.referencemode.ReferenceModeStorageKey
import arcs.core.type.Type

/**
 * [CapabilitiesResolver] is a factory class that creates [StorageKey]s based on the registered
 * providers and given [Capabilities].
 */
class CapabilitiesResolver(
  private val options: Options,
  private val factories: Map<String, StorageKeyFactory>,
  private val selector: FactorySelector = SimpleCapabilitiesSelector()
) {
  constructor(
    options: Options,
    factoriesList: List<StorageKeyFactory> = listOf(),
    selector: FactorySelector = SimpleCapabilitiesSelector()
  ) : this(options, CapabilitiesResolver.getFactoryMap(factoriesList), selector)

  /** Options used to construct [CapabilitiesResolver]. */
  data class Options(val arcId: ArcId)

  fun createStorageKey(capabilities: Capabilities, type: Type, handleId: String): StorageKey {
    val selectedFactories =
      factories.filterValues { it.supports(capabilities) }

    require(!selectedFactories.isEmpty()) {
      "Cannot create storage key for handle '$handleId' with capabilities $capabilities"
    }
    val factory = selector.select(selectedFactories.values)
    return createStorageKeyWithFactory(factory, type, handleId)
  }

  private fun createStorageKeyWithFactory(
    factory: StorageKeyFactory,
    type: Type,
    handleId: String
  ): StorageKey {
    val containerKey = factory.create(
      StorageKeyFactory.ContainerStorageKeyOptions(options.arcId, type.toSchema())
    )
    val containerChildKey = containerKey.childKeyForHandle(handleId)
    if (type is ReferenceType<*>) {
      return containerChildKey
    }
    val backingKey = factory.create(
      StorageKeyFactory.BackingStorageKeyOptions(options.arcId, type.toSchema())
    )
    // ReferenceModeStorageKeys in different drivers can cause problems with garbage collection.
    check(backingKey.protocol == containerKey.protocol) {
      "Backing and containers keys must use same protocol"
    }
    return ReferenceModeStorageKey(backingKey, containerChildKey)
  }

  /**
   * An interface for selecting a factory, if more than one are available for [Capabilities].
   */
  interface FactorySelector {
    fun select(factories: Collection<StorageKeyFactory>): StorageKeyFactory
  }

  /**
   * An implementation of a [FactorySelector] choosing a factory with a least restrictive max
   * [Capabilities] set.
   */
  class SimpleCapabilitiesSelector(
    val sortedProtocols: Array<String> = arrayOf("volatile", "ramdisk", "memdb", "db")
  ) : FactorySelector {
    override fun select(factories: Collection<StorageKeyFactory>): StorageKeyFactory {
      require(factories.isNotEmpty()) { "Cannot select from empty factories list" }
      val compareProtocol = compareBy { protocol: String ->
        val index = sortedProtocols.indexOf(protocol)
        if (index >= 0) {
          index
        } else sortedProtocols.size
      }

      return factories.reduce { acc, factory ->
        if (minOf(acc.protocol, factory.protocol, compareProtocol) == factory.protocol) {
          factory
        } else acc
      }
    }
  }

  companion object {
    private val defaultStorageKeyFactories = mutableMapOf<String, StorageKeyFactory>()

    fun registerStorageKeyFactory(factory: StorageKeyFactory) {
      require(defaultStorageKeyFactories[factory.protocol] == null) {
        "Storage key factory for '$factory.protocol' already registered"
      }
      defaultStorageKeyFactories[factory.protocol] = factory
    }

    fun reset() {
      defaultStorageKeyFactories.clear()
    }

    fun getFactoryMap(factoriesList: List<StorageKeyFactory>): Map<String, StorageKeyFactory> {
      require(factoriesList.distinctBy { it.protocol }.size == factoriesList.size) {
        "Storage keys protocol must be unique, but was: ${factoriesList.map {it.protocol}}."
      }
      val factories = factoriesList.associateBy { it.protocol }.toMutableMap()
      CapabilitiesResolver.defaultStorageKeyFactories.forEach { (protocol, factory) ->
        factories.getOrPut(protocol) { factory }
      }
      return factories
    }
  }
}
